test/src/main/java/org/apache/accumulo/test/functional/ZooMutatorIT.java

```
@@ -0,0 +1,119 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.accumulo.core.client.Accumulo;
+import org.apache.accumulo.core.clientImpl.ClientContext;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.fate.zookeeper.ZooReaderWriter;
+import org.apache.accumulo.harness.AccumuloClusterHarness;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.junit.Test;
+
+public class ZooMutatorIT extends AccumuloClusterHarness {
+  /**
+   * A simple stress test that looks for race conditions in
+   * {@link ZooReaderWriter#mutateOrCreate(String, byte[], org.apache.accumulo.fate.zookeeper.ZooReaderWriter.Mutator)}
+   */
+  @Test
+  public void concurrentMutatorTest() throws Exception {
+    try (var client = Accumulo.newClient().from(getClientProps()).build()) {
+
+      @SuppressWarnings("resource")
+      var context = (ClientContext) client;
+      var secret = cluster.getSiteConfiguration().get(Property.INSTANCE_SECRET);
+
+      ZooReaderWriter zk = new ZooReaderWriter(context.getZooKeepers(),
+          context.getZooKeepersSessionTimeOut(), secret);
+
+      var execServ = Executors.newFixedThreadPool(16);
+
+      String initialData = DigestUtils.sha1Hex("Accumulo Zookeeper Mutator test 1/4/21") + " 0";
+
+      // This map is used to ensure multiple threads do not successfully write the same value and no
+      // values are skipped. The hash in the value also verifies similar things in a different way.
+      ConcurrentHashMap<Integer,Integer> countCounts = new ConcurrentHashMap<>();
+
+      for (int i = 0; i < 16; i++) {
+        execServ.execute(() -> {
+          try {
+
+            int count = 0;
+            while (count < 200) {
+              byte[] val =
+                  zk.mutateOrCreate("/test-zm", initialData.getBytes(UTF_8), this::nextValue);
+              var nextCount = getCount(val);
+              assertTrue(nextCount > count);
```

This assertion can fail in threads, but since the executor has no uncaught exception handler, they just cause the threads to die, but the test passes anyway (with a messy stack trace printed to STDERR).

---

When switching to futures, I found this assertion was throwing an exception. Turns out the sanity checks in the test were not properly handling the initial insert into ZK and it was being silently ignored.